File: backend/src/schemas/runtime.py
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class ExecutionStepResult(BaseModel):
    task_id: str
    status: str
    output: Optional[str] = None
    tokens_used: int = 0
    duration_ms: Optional[int] = None
    model_name: Optional[str] = None
    worker_id: Optional[str] = None
    quota_status: Optional[str] = None
    is_handoff: bool = False


class GoalExecutionResult(BaseModel):
    goal_id: str
    status: str
    tasks_completed: int = 0
    tasks_failed: int = 0
    tasks_handoff: int = 0
    total_tokens_used: int = 0
    total_duration_ms: int = 0
    execution_log: List[Dict[str, Any]] = Field(default_factory=list)


class RuntimeStatusResponse(BaseModel):
    goal_id: str
    goal_title: str
    goal_status: str
    current_task_id: Optional[str] = None
    total_tasks: int = 0
    completed_tasks: int = 0
    failed_tasks: int = 0
    running_tasks: int = 0
    handoff_tasks: int = 0
    handoff_count: int = 0
    total_tokens_used: int = 0
    log_count: int = 0
    model_call_count: int = 0
    error_count: int = 0
    final_output: Optional[str] = None
    error_message: Optional[str] = None
    quota_status: Optional[str] = None

File: frontend/src/types/runtime.ts
export interface ExecutionStep {
  task_id: string;
  status: string;
  output?: string | null;
  tokens_used: number;
  duration_ms?: number | null;
  model_name?: string | null;
  worker_id?: string | null;
  quota_status?: string | null;
  is_handoff: boolean;
}

export interface GoalExecutionResult {
  goal_id: string;
  status: string;
  tasks_completed: number;
  tasks_failed: number;
  tasks_handoff: number;
  total_tokens_used: number;
  total_duration_ms: number;
  execution_log: ExecutionStep[];
}

export interface RuntimeStatusResponse {
  goal_id: string;
  goal_title: string;
  goal_status: string;
  current_task_id?: string | null;
  total_tasks: number;
  completed_tasks: number;
  failed_tasks: number;
  running_tasks: number;
  handoff_tasks: number;
  handoff_count: number;
  total_tokens_used: number;
  log_count: number;
  model_call_count: number;
  error_count: number;
  final_output?: string | null;
  error_message?: string | null;
  quota_status?: string | null;
  final_summary?: FinalSummary | null;
}

export interface FinalSummary {
  completed: string[];
  incomplete: string[];
  quality: {
    status: string;
    passed?: boolean | null;
    summary?: string | null;
    reviewer_model_id?: string | null;
  };
  risks: string[];
  models: Array<{ id: string; name: string; cost_level?: number | null }>;
  handoff_count: number;
  cost: { currency_estimate?: number | null; available: boolean; note: string };
}
